composer/callbacks/checkpoint_saver.py

```diff
@@ -0,0 +1,439 @@
+        write_mode = "w"
+    elif checkpoint_name.endswith(".tar.gz") or checkpoint_name.endswith(".tgz"):
+        write_mode = "w:gz"
+    elif checkpoint_name.endswith(".tar.bz2"):
+        write_mode = "w:bz2"
+    elif checkpoint_name.endswith(".tar.lzma"):
+        write_mode = "w:xz"
+    else:
+        raise ValueError(
+            textwrap.dedent(f"""\
+            Checkpoint name ({checkpoint_name}) has an unsupported file extension.
+            Must be one of .pt, .tar, .tgz, .tar.gz, .tar.bz2, or .tar.lzma."""))
+
+    return write_mode
+
+
+def checkpoint_periodically(interval: Union[str, int, Time]) -> Callable[[State, Event], bool]:
+    """Helper function to create a checkpoint scheduler according to a specified interval.
+
+    Args:
+        interval (Union[str, int, Time]): The interval describing how often checkpoints should be
+            saved. If an integer, it will be assumed to be :attr:`TimeUnit.EPOCH`.
+            Otherwise, the unit must be in :attr:`TimeUnit.EPOCH` or :attr:`TimeUnit.BATCH`.
+
+            Checkpoints will be saved every ``n`` batches or epochs (depending on the unit),
+            and at the end of training.
+
+    Returns:
+        Callable[[State, Event], bool]: A function that can be passed as the ``should_save``
+            argument into :class:`CheckpointSaver`.
+    """
+    if isinstance(interval, str):
+        interval = Time.from_timestring(interval)
+    if isinstance(interval, int):
+        interval = Time(interval, TimeUnit.EPOCH)
+
+    if interval.unit == TimeUnit.EPOCH:
+        save_event = Event.EPOCH_CHECKPOINT
+    elif interval.unit == TimeUnit.BATCH:
+        save_event = Event.BATCH_CHECKPOINT
+    else:
+        raise NotImplementedError(
+            f"Unknown checkpointing interval: {interval.unit}. Must be TimeUnit.EPOCH or TimeUnit.BATCH.")
+
+    last_checkpoint_batch = None
+
+    def should_save(state: State, event: Event):
+        nonlocal last_checkpoint_batch
+        if state.get_elapsed_duration() >= 1.0:
+            # if doing batch-wise checkpointing, and we saved a checkpoint at the batch_checkpoint event
+            # right before the epoch_checkpoint event, do not save another checkpoint at the epoch_checkpoint
+            # event if the batch count didn't increase.
+            if state.timer.batch != last_checkpoint_batch:
+                last_checkpoint_batch = state.timer.batch
+                return True
+
+        if event == save_event:
+            if save_event == Event.EPOCH_CHECKPOINT and int(state.timer.epoch) % int(interval) == 0:
+                last_checkpoint_batch = state.timer.batch
+                return True
+            if save_event == Event.BATCH_CHECKPOINT and int(state.timer.batch) % int(interval) == 0:
+                last_checkpoint_batch = state.timer.batch
+                return True
+
+        return False
+
+    return should_save
+
+
+class CheckpointSaver(Callback):
```

Comment:
Currently it's difficult for someone to extend this callback because its unclear in this file which are helper functions for saving the checkpoint, and which are unique to CheckpointSaver. I would suggest we aim to make `CheckpointSaver` as minimal as possible, as move the save_checkpoint code and associated helpers to a different file. As suggested in the design, we can make it a static method to the trainer.

Reply:
See `composer.utils.checkpoint:save_checkpoint`